### ISSUE TITLE
why train with (img - 128)/128, but get hard example with (img - 128) only?

### ISSUE THREAD

**Issue by bpf476499866** [CongWeilin/mtcnn-caffe#40]
Hi! Sorry to bother you again.

Noticing the model is trained with image normalized to 【(img - 128)/128】, range [-1, 1]
And 【demo/test.py, line 23】do use 【caffe_img = (img.copy()-127.5)/128】 as net input;

But 【prepare_data/gen_12net_hard_example.py, line 22】, only 【caffe_img = img.copy()-128】as input,
I guess ... this line has just been forgotten during your code modifying, isn't it?

**Comment by haichaoyu** [comment]
I am wondering whether the code could reproduce the results in the paper.

**Comment by Rogerluojie** [comment]
What is the function of this file：gen_12net_hard_example.py？

**Comment by bpf476499866** [comment]
@Rogerluojie 
let all training set go pass through the 12-net with a trained 12-net model, and harvest rects detected but are not real faces (false positive). these rects are what we called "hard example", witch will be used as negative samples to fine-tune the existing 12-net model, so as to promote the model's robustness against those none-face part of a image.

**Comment by Rogerluojie** [comment]
@bpf476499866 
OK,THX! but I have doubt that in this file(gen_12net_hard_example.py)we generate three file (neg_24.txt,part_24.txt,pos_24.txt )the same as the other file(gen_24net_data.py)do?

**Comment by bpf476499866** [comment]
@Rogerluojie 
em..... The 24-net also need to be trained twice, yet the first-time training will no longer use random sample but the hard samples from 12net, in order to 【boost】the hole cascade. 
So the training data in first round need to be same as 12-net's finetune data, the only different is that the size should be resized into 24*24 to fit 24-net, but not 12*12.
If I did not make it clear enough, there is another re-implement whose read-me is more friendly.

**Comment by Rogerluojie** [comment]
@bpf476499866 
thank you for your detailed interpretation！new to study the training so have many problem and sorry to disturb you.
